tests/js/scroll/scroll-bidirectional-tests.js

```diff
@@ -0,0 +1,249 @@
+            gamepad.tests.windowObject.navigator.getGamepads = function () {
+                return gamepad.tests.utils.bidirectional.oneAxis(1);
+            };
+            $(window).scrollTop(0);
+            jqUnit.assertEquals("The vertical scroll position should be at the top of the webpage.", 0, $(window).scrollTop());
+
+            var testNavigator = gamepad.inputMapper({
+                windowObject: gamepad.tests.windowObject,
+                frequency: gamepad.tests.frequency
+            });
+            jqUnit.assertValue("The Gamepad Navigator should be instantiated.", testNavigator);
+
+            // Trigger gamepadconnected event and start scrolling vertically.
+            testNavigator.events.onGamepadConnected.fire();
+            var yPosition = null;
+            gamepad.tests.intervalLoop = setInterval(function () {
+                if (gamepad.tests.windowObject.count === 4) {
+                    yPosition = $(window).scrollTop();
+                    jqUnit.assertNotEquals("The page should have scrolled down.", 0, yPosition);
+                }
+                else if (gamepad.tests.windowObject.count === 0) {
+                    var isScrolledUp = $(window).scrollTop() < yPosition ? true : false;
+                    jqUnit.assertTrue("The page should have scrolled up.", isScrolledUp);
+
+                    testNavigator.events.onGamepadDisconnected.fire();
+                    clearInterval(gamepad.tests.intervalLoop);
+                    testNavigator.destroy();
+                    jqUnit.assertTrue("The instance of the gamepad navigator should be destroyed.", fluid.isDestroyed(testNavigator));
+                    jqUnit.start();
+                }
+            }, gamepad.tests.frequency);
+        });
+
+        jqUnit.asyncTest("Scroll diagonally towards the bottom right", function () {
+            jqUnit.expect(4);
+
+            // No of times the gamepad mock should be returned.
+            gamepad.tests.windowObject.count = 5;
+            gamepad.tests.windowObject.navigator.getGamepads = function () {
+                return gamepad.tests.utils.bidirectional.diagonalAxes({
+                    "0": 1,
+                    "1": 1
+                });
+            };
+            $(window).scrollTop(0);
+            $(window).scrollLeft(0);
+            var isTopLeft = ($(window).scrollTop() === 0) && ($(window).scrollLeft() === 0);
+            jqUnit.assertTrue("The scroll position should be at the top-left of the webpage.", isTopLeft);
+
+            var testNavigator = gamepad.inputMapper({
+                windowObject: gamepad.tests.windowObject,
+                frequency: gamepad.tests.frequency
+            });
+            jqUnit.assertValue("The Gamepad Navigator should be instantiated.", testNavigator);
+
+            // Trigger gamepadconnected event and start scrolling towards bottom-right corner.
+            testNavigator.events.onGamepadConnected.fire();
```

Comment:
I see the pattern you're using, which is to have a gamepad in a permanent state, then connect it, then eventually disconnect it and look at the change.  I'm thinking that we need a test version of whatever component is polling the gamepads that can be manually instructed to poll the gamepads.   We'd also need a gamepad whose state can be updated.  That way you can much more cleanly establish the sequence of events in these tests, i.e.:

1. Scroll the page to the top.
2. Confirm the initial state.
3. Connect a gamepad with no buttons.
4. Instruct the test component to poll the gamepads once.
5. Confirm that the scrolling hasn't changed.
6. Update the gamepad to hold down a control.
7. Instruct the test component to poll the gamepads once.
8. Confirm the scrolling has changed.

You would only need to test the recurring polling mechanism (setInterval) in your end-to-end tests.

Reply:
I think I'm fine with the slightly different approach you used.  There is a small risk that the tests will fail if you connect or disconnect a gamepad while they're running, but I'd rather revisit if we encounter problems doing things this way.  I do think that eventually you'll want a more convenient pattern than writing out the replacement functions, but at least what you have now it does make it fairly clear what you're doing and avoids potential race conditions with competing intervals and timeouts.